micronota/parsers/cmscan.py

```diff
@@ -0,0 +1,413 @@
+                           'SEQUENCE_NAME',
+                           'SEQUENCE_ACCESSION',
+                           'MODEL_NAME',
+                           'MODEL_ACCESSION']:
+                    continue
+                attributes[key] = fields[_COLUMNS[key]['position']]
+
+#           cmscan works on multiple sequence in one FASTA file. We want to
+#           yield a separate object for each sequence, thus we create a new
+#           one whenever the ID changes
+            if (currentsequence != attributes['SEQUENCE_NAME'] and
+                    currentsequence is not False):
+                yield annotations
+                annotations = {}
+
+#           a metadata interval is made out of 'frozen dictionary' aka
+#           Features object, where the its hashable value constitutes the
+#           key and the value is a list of tuples, aka intervals
+            annotations[Feature(attributes)] = [(hitStart, hitEnd)]
+#           store current sequence id for the next iteration
+            currentsequence = attributes['SEQUENCE_NAME']
+
+    yield annotations
+
+
+def _writeheader(fh):
+    """ Print a header for cmscan hits. It is similar to 'CMscan --tblout' but
+        not identical, since we do not check for longest occurring names, but
+        use fixed field width. """
+
+    # write file header
+    lineCaption = '#'
+    lineDelimeter = '#'
+    for key in _orderedKeys:
+        length = _COLUMNS[key]['width']
+        if _COLUMNS[key]['position'] == 0:
+            length = length-1
+        lineCaption = lineCaption + _printField(
+               _COLUMNS[key], _COLUMNS[key]['caption'], length)
+        lineDelimeter = lineDelimeter + ('-' * length)
+        if _COLUMNS[key]['position']+1 != len(_COLUMNS):
+            lineCaption = lineCaption + " "
+            lineDelimeter = lineDelimeter + " "
+    fh.write(lineCaption + "\n")
+    fh.write(lineDelimeter + "\n")
+
+
+@cmscan.writer(IntervalMetadata)
+def _IntervalMetadata_to_cmscan(obj, fh):
+    """ Prints IntervalMetadata object in a format similar to CMscan --tblout.
+        It is not identical, since ordering is arbitrary here. Furthermore,
+        field width might vary. """
+#   write data
+    for hit, interval in obj.features.items():
+        if len(interval) != 1:
+            raise CmscanFormatError(
+             "The cmscan format allows only one interval per IntervalMetadata "
+             "object. Cannot continue writing output.")
+        lineData = ''
+        for key in _orderedKeys:
+            if key == 'SEQUENCE_START_POSITION':
+                if hit['STRAND'] == '+':
+                    value = interval[0][0].__str__()
+                else:
+                    value = interval[0][1].__str__()
+            elif key == 'SEQUENCE_END_POSITION':
+                if hit['STRAND'] == '+':
+                    value = interval[0][1].__str__()
+                else:
+                    value = interval[0][0].__str__()
```

Comment:
it now iterates over all intervals and prints them each in one line (annotation data will be identical, except of start and end positions)